File: mcp-servers/alp-mcp-server.py
"""
AgentHub ALP — MCP Server 示例 (Python)
作为 Skills 与 PolarDB 之间的桥梁，通过 MCP 协议暴露数据库操作能力。

运行方式:
  python alp-mcp-server.py --port 8000

依赖:
  pip install mcp polar-connector psycopg2-binary
"""

import json
import os
from mcp import MCPServer, Tool

# ── ALP Skills 工具定义 ──

TOOLS = [
    Tool(
        name="data-query",
        description="查询 PolarDB 结构化数据（绑定 C09 data Worker）",
        input_schema={
            "type": "object",
            "properties": {
                "sql": {"type": "string", "description": "SQL 查询语句"},
                "params": {"type": "object", "description": "查询参数"},
                "limit": {"type": "integer", "default": 100},
            },
            "required": ["sql"],
        },
        handler=lambda ctx, args: query_database(args),
    ),
    Tool(
        name="vector-store",
        description="将内容向量化存储（绑定 C07 memory Worker）",
        input_schema={
            "type": "object",
            "properties": {
                "content": {"type": "string"},
                "metadata": {"type": "object"},
                "memory_type": {
                    "type": "string",
                    "enum": ["short-term", "long-term"],
                },
            },
            "required": ["content"],
        },
        handler=lambda ctx, args: store_vector(args),
    ),
    Tool(
        name="similarity-search",
        description="向量相似性搜索（绑定 C07 memory Worker）",
        input_schema={
            "type": "object",
            "properties": {
                "query": {"type": "string"},
                "top_k": {"type": "integer", "default": 5},
                "threshold": {"type": "number", "default": 0.7},
            },
            "required": ["query"],
        },
        handler=lambda ctx, args: search_similar(args),
    ),
    Tool(
        name="knowledge-retrieval",
        description="RAG 知识检索（绑定 C08 knowledge Worker）",
        input_schema={
            "type": "object",
            "properties": {
                "query": {"type": "string"},
                "top_k": {"type": "integer", "default": 3},
            },
            "required": ["query"],
        },
        handler=lambda ctx, args: retrieve_knowledge(args),
    ),
    Tool(
        name="report-generator",
        description="生成结构化报告（绑定 A09 generate Worker）",
        input_schema={
            "type": "object",
            "properties": {
                "title": {"type": "string"},
                "sections": {"type": "array"},
                "format": {
                    "type": "string",
                    "enum": ["markdown", "html", "json"],
                    "default": "markdown",
                },
            },
            "required": ["title", "sections"],
        },
        handler=lambda ctx, args: generate_report(args),
    ),
    Tool(
        name="statistical-analysis",
        description="数据统计分析（绑定 A04 analyze Worker）",
        input_schema={
            "type": "object",
            "properties": {
                "data": {"type": "array"},
                "analysis_type": {
                    "type": "string",
                    "enum": ["descriptive", "trend", "correlation", "distribution"],
                },
            },
            "required": ["data", "analysis_type"],
        },
        handler=lambda ctx, args: analyze_data(args),
    ),
]


# ── 处理函数 ──

def query_database(args):
    sql = args.get("sql")
    params = args.get("params", {})
    limit = args.get("limit", 100)

    # 实际实现：连接到 PolarDB 执行 SQL
    # conn = psycopg2.connect(os.environ["POLARDB_DSN"])
    # cur = conn.cursor()
    # cur.execute(sql, params)
    # rows = cur.fetchmany(limit)
    # columns = [desc[0] for desc in cur.description]

    return {
        "columns": ["id", "name", "value", "date"],
        "rows": [
            {"id": 1, "name": "示例数据", "value": 100, "date": "2024-01-01"},
            {"id": 2, "name": "示例数据", "value": 200, "date": "2024-02-01"},
        ],
        "row_count": 2,
        "_note": "This is a stub. Replace with actual PolarDB connection.",
    }


def store_vector(args):
    content = args["content"]
    metadata = args.get("metadata", {})
    memory_type = args.get("memory_type", "long-term")

    # 实际实现：调用 embedding API + 写入 PolarDB pgvector
    # embedding = openai.Embedding.create(input=content, model="text-embedding-3-large")
    # cur.execute("INSERT INTO alp_memories (content, embedding, type, metadata) VALUES (%s, %s, %s, %s)",
    #             (content, embedding, memory_type, json.dumps(metadata)))

    return {
        "memory_id": "mem-uuid-here",
        "embedding_dim": 1536,
        "status": "stored",
    }


def search_similar(args):
    query = args["query"]
    top_k = args.get("top_k", 5)
    threshold = args.get("threshold", 0.7)

    # 实际实现：向量检索
    # embedding = openai.Embedding.create(input=query, ...)
    # cur.execute("SELECT content, metadata, 1 - (embedding <=> %s) AS score FROM alp_memories WHERE 1 - (embedding <=> %s) > %s ORDER BY score DESC LIMIT %s",
    #             (embedding, embedding, threshold, top_k))

    return {
        "results": [],
        "total_count": 0,
    }


def retrieve_knowledge(args):
    query = args["query"]
    top_k = args.get("top_k", 3)

    return {
        "chunks": [
            {
                "content": f"与 '{query}' 相关的知识片段",
                "source": "knowledge-base",
                "relevance": 0.95,
            }
        ],
        "relevance_scores": [0.95],
    }


def generate_report(args):
    title = args["title"]
    sections = args["sections"]
    fmt = args.get("format", "markdown")

    report = f"# {title}\n\n"
    for section in sections:
        if isinstance(section, str):
            report += f"## {section}\n\n内容区域\n\n"
        elif isinstance(section, dict):
            report += f"## {section.get('heading', '')}\n\n{section.get('content', '')}\n\n"

    return {
        "report": report,
        "format": fmt,
        "length": len(report),
    }


def analyze_data(args):
    data = args["data"]
    analysis_type = args["analysis_type"]

    import statistics

    values = [d.get("value", 0) if isinstance(d, dict) else d for d in data]

    result = {"data_points": len(data), "analysis_type": analysis_type}

    if analysis_type == "descriptive":
        result.update(
            {
                "mean": statistics.mean(values),
                "median": statistics.median(values),
                "min": min(values),
                "max": max(values),
                "stdev": statistics.stdev(values) if len(values) > 1 else 0,
            }
        )
    elif analysis_type == "trend":
        result.update(
            {"direction": "up" if values[-1] > values[0] else "down", "change_pct": 0}
        )

    return {"statistics": result, "insights": [f"共 {len(data)} 条数据，分析完成"], "visualization": None}


# ── 启动 MCP Server ──

if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="ALP MCP Server")
    parser.add_argument("--port", type=int, default=8000, help="监听端口")
    args = parser.parse_args()

    server = MCPServer(
        name="alp-mcp-server",
        version="1.0.0",
        tools=TOOLS,
        description="AgentHub ALP — Skills MCP Server，为 60 个智能体提供真实操作能力",
    )
    server.run(host="0.0.0.0", port=args.port)

File: mcp-servers/alp-mcp-server.ts
/**
 * AgentHub ALP — MCP Server 示例 (TypeScript/Node.js)
 * 
 * 作为 Skills 与 PolarDB 之间的桥梁，通过 MCP 协议暴露 ALP 智能体操作能力。
 * 运行方式:
 *   npx tsx alp-mcp-server.ts --port 8000
 * 
 * 依赖:
 *   npm install @modelcontextprotocol/sdk polar-connector pg
 */

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
} from "@modelcontextprotocol/sdk/types.js";
import { Pool } from "pg";

// PolarDB 连接池
let pool: Pool;

function getPool(): Pool {
  if (!pool) {
    pool = new Pool({
      connectionString: process.env.POLARDB_DSN,
      max: 10,
      idleTimeoutMillis: 30000,
    });
  }
  return pool;
}

// ── ALP Skills 工具定义 ──

const TOOLS = [
  {
    name: "data-query",
    description: "查询 PolarDB 结构化数据（绑定 C09 data Worker）",
    inputSchema: {
      type: "object",
      properties: {
        sql: { type: "string", description: "SQL 查询语句" },
        params: { type: "object", description: "查询参数" },
        limit: { type: "integer", default: 100 },
      },
      required: ["sql"],
    },
  },
  {
    name: "vector-store",
    description: "将内容向量化存储（绑定 C07 memory Worker）",
    inputSchema: {
      type: "object",
      properties: {
        content: { type: "string" },
        metadata: { type: "object" },
        memoryType: {
          type: "string",
          enum: ["short-term", "long-term"],
        },
      },
      required: ["content"],
    },
  },
  {
    name: "similarity-search",
    description: "向量相似性搜索（绑定 C07 memory Worker）",
    inputSchema: {
      type: "object",
      properties: {
        query: { type: "string" },
        topK: { type: "integer", default: 5 },
        threshold: { type: "number", default: 0.7 },
      },
      required: ["query"],
    },
  },
  {
    name: "knowledge-retrieval",
    description: "RAG 知识检索（绑定 C08 knowledge Worker）",
    inputSchema: {
      type: "object",
      properties: {
        query: { type: "string" },
        topK: { type: "integer", default: 3 },
      },
      required: ["query"],
    },
  },
  {
    name: "statistical-analysis",
    description: "数据统计分析（绑定 A04 analyze Worker）",
    inputSchema: {
      type: "object",
      properties: {
        data: { type: "array" },
        analysisType: {
          type: "string",
          enum: ["descriptive", "trend", "correlation", "distribution"],
        },
      },
      required: ["data", "analysisType"],
    },
  },
  {
    name: "report-generator",
    description: "生成结构化报告（绑定 A09 generate Worker）",
    inputSchema: {
      type: "object",
      properties: {
        title: { type: "string" },
        sections: { type: "array" },
        format: {
          type: "string",
          enum: ["markdown", "html", "json"],
          default: "markdown",
        },
      },
      required: ["title", "sections"],
    },
  },
];

const server = new Server(
  {
    name: "alp-mcp-server",
    version: "1.0.0",
  },
  {
    capabilities: {
      tools: {},
    },
  }
);

// ── 处理函数 ──

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: TOOLS,
}));

server.setRequestHandler(CallToolRequestSchema, async (request) => {
  const { name, arguments: args } = request.params;

  switch (name) {
    case "data-query":
      return await handleDataQuery(args);
    case "vector-store":
      return handleVectorStore(args);
    case "similarity-search":
      return handleSimilaritySearch(args);
    case "knowledge-retrieval":
      return handleKnowledgeRetrieval(args);
    case "statistical-analysis":
      return handleStatisticalAnalysis(args);
    case "report-generator":
      return handleReportGenerator(args);
    default:
      throw new Error(`Unknown tool: ${name}`);
  }
});

async function handleDataQuery(args: any) {
  const sql = args.sql;
  const limit = args.limit || 100;

  try {
    const client = await getPool().connect();
    const result = await client.query(sql);
    client.release();

    return {
      content: [
        {
          type: "text",
          text: JSON.stringify({
            columns: result.fields.map((f: any) => f.name),
            rows: result.rows.slice(0, limit),
            rowCount: result.rowCount,
          }),
        },
      ],
    };
  } catch (err: any) {
    return {
      content: [{ type: "text", text: `查询失败: ${err.message}` }],
      isError: true,
    };
  }
}

function handleVectorStore(args: any) {
  // 调用 embedding API 并写入 PolarDB pgvector
  return {
    content: [
      {
        type: "text",
        text: JSON.stringify({
          memoryId: crypto.randomUUID(),
          status: "stored",
        }),
      },
    ],
  };
}

function handleSimilaritySearch(args: any) {
  return {
    content: [
      {
        type: "text",
        text: JSON.stringify({
          results: [],
          totalCount: 0,
        }),
      },
    ],
  };
}

function handleKnowledgeRetrieval(args: any) {
  return {
    content: [
      {
        type: "text",
        text: JSON.stringify({
          chunks: [
            {
              content: `与 '${args.query || ""}' 相关的知识片段`,
              source: "knowledge-base",
              relevance: 0.95,
            },
          ],
        }),
      },
    ],
  };
}

function handleStatisticalAnalysis(args: any) {
  const data = args.data || [];
  const analysisType = args.analysisType || "descriptive";
  const values = data.map((d: any) => (typeof d === "number" ? d : d.value || 0));

  const n = values.length;
  const mean = values.reduce((a: number, b: number) => a + b, 0) / n;
  const sorted = [...values].sort((a, b) => a - b);

  return {
    content: [
      {
        type: "text",
        text: JSON.stringify({
          statistics: {
            mean,
            median: sorted[Math.floor(n / 2)],
            min: sorted[0],
            max: sorted[n - 1],
            count: n,
          },
          insights: [`${n} 条数据 ${analysisType} 分析完成`],
        }),
      },
    ],
  };
}

function handleReportGenerator(args: any) {
  const title = args.title;
  const sections = args.sections || [];

  let report = `# ${title}\n\n`;
  for (const section of sections) {
    if (typeof section === "string") {
      report += `## ${section}\n\n内容区域\n\n`;
    } else {
      report += `## ${section.heading || ""}\n\n${section.content || ""}\n\n`;
    }
  }

  return {
    content: [
      {
        type: "text",
        text: JSON.stringify({ report, format: args.format || "markdown", length: report.length }),
      },
    ],
  };
}

// ── 启动 ──

async function main() {
  const transport = new StdioServerTransport();
  await server.connect(transport);
  console.error("ALP MCP Server running on stdio");
}

main().catch(console.error);
